File: 02_Stack/05_Car_Fleet/853_Car_Fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {

        stack<double> res;

        vector<pair<int, int>> cars;
        for (int i = 0; i < position.size(); i++){
            cars.push_back({position[i], speed[i]});
        }

        sort(cars.rbegin(), cars.rend());


        for (int i=0; i<cars.size(); i++){
            double time = double(target - cars[i].first) / cars[i].second;
            res.push(time);
            cout << time;
            if (res.size() >= 2){
                double current_car = res.top();
                res.pop();
                double prev_car = res.top();
                if (current_car > prev_car){
                    res.push(current_car);
                }
            }
        }

        return res.size();

    }
};

File: 02_Stack/05_Car_Fleet/853_Car_Fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:

        cars = [(pos, speed) for pos, speed in zip(position, speed)]
        cars.sort(reverse=True)
        res = []

        for pos, speed in cars:
            res.append((target - pos) / speed)
            if len(res) >= 2 and res[-1] <= res[-2]:
                res.pop()

        return len(res)